File: 6_kyu/Playing_with_digits.lua
-- Returns array with decimal digits of x.
-- Index 1 = lowest significant digit
local function digits(x)
  local d = {}
  while x > 0 do
    table.insert(d, x % 10)
    x = x // 10
  end
  return d
end

return {
  dig_pow = function(n,pow)
    local d = digits(n)
    local sum = 0
    for i = 1,#d do
      sum = sum + d[i] ^ (#d+pow-i)
    end
    local div = sum / n
    if div % 1 == 0 then
      return div
    else
      return -1
    end
  end
  }

File: 6_kyu/Playing_with_digits.kt
/** Power function for integers */
fun Int.pow(exp: Int): Long {
    return this.toBigInteger().pow(exp).toLong()
}

/**
 * Sum decimal digits with power.
 *
 * @param x Calculate the digit powerd sum of this int.
 * @param p The exponent of the most significant digit.
 * @return sum of all decimal digits with power p,p+1,..
 */
fun digitSum(x: Int, p: Int): Long {
    var digits = IntArray(32)
    assert( 10.pow(digits.size) > Int.MAX_VALUE.toLong() )
    var nDigits = 0
    var rest = x
    while (rest > 0) {
        digits[nDigits] = rest % 10
        rest /= 10
        nDigits++
    }
    var sum: Long = 0
    for (i in 0..nDigits - 1) {
        sum += digits[i].pow(p + nDigits - i - 1)
    }
    return sum
}

/**
 * Calculates funny integer.
 *
 * @param n Integer to analyse.
 * @param p exponent of the most significant digit in the sum.
 * @return k with k*n = powered decimal digit sum of n.
 */
fun digPow(n: Int, p: Int): Int {
    val s = digitSum(n, p)
    return if (s % n == 0L) (s / n).toInt() else -1
}
